app/Services/Validation/ApplicationTimelineValidator.php

```diff
@@ -0,0 +1,421 @@
+            ->pluck('id');
+
+        $validator = Validator::make($application->toArray(), $this->basicInfoSimpleRules());
+
+        // Conditionally check for 'language_test_confirmed' if the
+        // closure returns true.
+        $validator->sometimes(
+            'language_test_confirmed',
+            'required|boolean|accepted',
+            function ($input) use ($langRequirement, $langTestRequirement) {
+                return in_array($langRequirement, $langTestRequirement->toArray());
+            }
+        );
+
+        return $validator;
+    }
+
+    /**
+     * Helper function to return completeness for the Basic Info step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return boolean
+     */
+    public function basicsComplete(JobApplication $application)
+    {
+        $validator = $this->basicsValidator($application);
+        return $validator->passes();
+    }
+
+    /* End Step 1 --------------------------------------------------------------------------------------------------- */
+
+    /* Step 2 ------------------------------------------------------------------------------------------------------- */
+
+    /**
+     * Validator instance for the Experience step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return \Illuminate\Contracts\Validation\Validator
+     */
+    public function experienceValidator(JobApplication $application)
+    {
+        $application->loadMissing('job_poster', 'job_poster.criteria');
+
+        $essentialCriteriaType = CriteriaType::where('name', 'essential')->first()->id;
+
+        $jobCriteria = $application->job_poster->criteria;
+        $requiredCriteria = $jobCriteria->filter(function ($criterion) use ($essentialCriteriaType) {
+            return $criterion->criteria_type_id === $essentialCriteriaType;
+        });
+
+        $experienceSkillRules = array();
+
+        foreach ($requiredCriteria as $criterion) {
+            // Validate that every essential skill has a corresponding experience.
+            $experienceSkillRules[] = [
+                'skill_id' => new ContainsArrayWithAttributeRule('skill_id', $criterion->skill_id)
+            ];
+        }
+
+        // Get all Experiences belonging to the current application, that are assigned to required Criteria.
+        $experiences = ExperienceSkill::whereHasMorph(
+            'experience',
+            '*',
+            function (Builder $query) use ($application): void {
+                $query->where([
+                    ['experienceable_type', 'application'],
+                    ['experienceable_id', $application->id]
+                ]);
+            }
+        )
+        ->whereIn('skill_id', $requiredCriteria->pluck('skill_id')->all())
+        ->get();
+
+        $validator = Validator::make($experiences->toArray(), $experienceSkillRules);
+
+        return $validator;
```

Comment:
So this bit:
```php
->whereIn('skill_id', $requiredCriteria->pluck('skill_id')->all())
```
should be limiting the query to only bring back relevant experiences (if they exist). We can definitely add an orderBy to ensure they both line up, if that helps

Reply:
Maybe the validator itself is the issue, I had assumed the `ContainsArrayWithAttributeRule` would validate as long as the `$experiences` array had one of each of a `skill_id` of a given number. If it's not working that way, then I guess I need to adjust the validation logic.

Reply:
Ah okay cool good to know. However, ya we'll need to refactor it cause it, like I said earlier it still passes when `$experiences` are empty. Actually, as long as the `ContainsArrayWithAttributeRule` rules never lines up with an incorrect `skill_id` then it passes. Even if the remaining rules aren't met.

Reply:
I think the ContainsArrayWithAttributeRule doesn't work as you expect. The way I read it, this will check that the 'skill_id' field of the validated object is an array that also has a 'skill_id' field (with the correct value). I'm also not sure how this will work when the validated object is itself an array.